### ISSUE TITLE
Bug when Router appears, disappears, then appears again

### ISSUE THREAD

**Issue by lrettig** [aksonov/react-native-router-flux#221]
If I switch displaying of the top level router on or off inside my `render()` method--so that it first appears, then disappears, then appears again--then when it comes back, I get this error when calling any route actions: "Cannot read property 'push' of undefined" (see stack trace below).

The use case here is that I currently have a few components, e.g. full screen modals that cover the nav bar, which live outside of the top-level Router. I suppose I might be able to move these inside the Router, but this still seems like an odd bug that should be fixed.

E.g.:

```
  render() {
    return this.state.mode ? <SomeOtherComponent/> : (
      <View style={{flex: 1}}>
        <Router>
          <Route name="screen1" .../>
          <Route name="screen2" .../>
        </Router>
      </View>
    );
  }
```

Full sample code here: https://github.com/lrettig/router-onoff-bug-example

![image](https://cloud.githubusercontent.com/assets/3316532/13130772/97184f26-d5b4-11e5-8a13-b59058a2c0f5.png)

**Comment by aksonov** [comment]
You should pass 'route' prop to children Router.

**Comment by lrettig** [comment]
Thanks for the tip! But I only have one `Router`. What do you mean?

**Comment by aksonov** [comment]
Have you tried latest npm version?

**Comment by lrettig** [comment]
The simple example I put together appears to work with the latest npm version--sorry for not trying that before. But upgrading my code is producing a new warning, and I'm still getting an error in the same place. I will keep debugging, but in case either of this ring a bell for you or you have any tips.

Here's the warning (I haven't changed my code; this is new with the latest npm version, I didn't get a warning before):

```
Possible Unhandled Promise Rejection (id: 0):
Cannot read property 'delegate' of null
TypeError: Cannot read property 'delegate' of null
    at ExRouter.componentWillUnmount (http://localhost:8081/index.ios.bundle?platform=ios&dev=true:72287:42)
    at ReactCompositeComponentMixin.unmountComponent (http://localhost:8081/index.ios.bundle?platform=ios&dev=true:15041:6)
    at Object.ReactReconciler.unmountComponent (http://localhost:8081/index.ios.bundle?platform=ios&dev=true:13509:18)
    at ReactCompositeComponentMixin.unmountComponent (http://localhost:8081/index.ios.bundle?platform=ios&dev=true:15044:17)
    at Object.ReactReconciler.unmountComponent (http://localhost:8081/index.ios.bundle?platform=ios&dev=true:13509:18)
    at Object.ReactChildReconciler.unmountChildren (http://localhost:8081/index.ios.bundle?platform=ios&dev=true:26600:17)
    at ReactMultiChild.Mixin.unmountChildren (http://localhost:8081/index.ios.bundle?platform=ios&dev=true:26386:22)
    at ReactNativeBaseComponent.Mixin.unmountComponent (http://localhost:8081/index.ios.bundle?platform=ios&dev=true:25827:6)
    at Object.ReactReconciler.unmountComponent (http://localhost:8081/index.ios.bundle?platform=ios&dev=true:13509:18)
    at ReactCompositeComponentMixin.unmountComponent (http://localhost:8081/index.ios.bundle?platform=ios&dev=true:15044:17)
```

Here's the error: "No current router is set":

![image](https://cloud.githubusercontent.com/assets/3316532/13149088/18daa986-d62e-11e5-9f86-9f4764fca8e2.png)

**Comment by lrettig** [comment]
My code uses tab bars and does have a hierarchy of `Routers`. What do I pass in the `route` prop to the child Router?

**Comment by lrettig** [comment]
I was able to work around this by moving the component in question _inside_ the top-level `Router`. I still think there are reasons you might sometimes want things to live outside the Router, and control their presentation via state, e.g., to easily present a modal that covers the entire screen (and leaves the view below in place), or if you want to swap a screen without any animation/transition, or if you want multiple `Router` hierarchies for different parts of your app, or if you simply prefer to keep state inside the state variables for part of your app.

But I may be misunderstanding these cases.

Feel free to close this if you like. Thanks again.

**Comment by MITDelian** [comment]
I've been running into this issue as well

**Comment by MITDelian** [comment]
Somehow the router is unmounted which I don't understand how it is possible

**Comment by mmazzarolo** [comment]
Maybe this can be useful:
Take a look at [this comment](https://github.com/aksonov/react-native-router-flux/issues/92#issuecomment-182322505). I had the same problem in nested routers and fixed it following the hints of the issue #172.

**Comment by aksonov** [comment]
Closing for now. Some fixes was made, please check latest version
